code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/request/handler/GenericWhitelistedProxy.java

```diff
@@ -0,0 +1,187 @@
+            redirect = protocol + "://" + redirect;
+        }
+
+        URI url = URI.create(redirect);
+        String host = url.getHost();
+
+        if (!isWhitelisted(StringUtils.strip(host, "/"))) {
+            throw new ClientVisibleException(ResponseCodes.FORBIDDEN);
+        }
+
+        Request temp;
+        String method = request.getServletContext().getRequest().getMethod();
+        switch (method) {
+        case "POST":
+            temp = Request.Post(redirect);
+            break;
+        case "GET":
+            temp = Request.Get(redirect);
+            break;
+        case "PUT":
+            temp = Request.Put(redirect);
+            break;
+        case "DELETE":
+            temp = Request.Delete(redirect);
+            break;
+        case "HEAD":
+            temp = Request.Head(redirect);
+            break;
+        default:
+            throw new ClientVisibleException(ResponseCodes.BAD_REQUEST, "Invalid method", "The method " + method + " is not supported", null);
+        }
+
+        @SuppressWarnings("unchecked")
+        List<String> restrictedHeaders = Collections.list(request.getServletContext().getRequest().getHeaders("X-API-HEADERS-RESTRICT"));
+
+        if (restrictedHeaders.size() == 1 && StringUtils.contains(restrictedHeaders.get(0), ",")) {
+            String[] headers = StringUtils.split(restrictedHeaders.get(0), ",");
+            restrictedHeaders.remove(0);
+            for (String header : headers) {
+                restrictedHeaders.add(StringUtils.trim(header));
+            }
+        }
+
+        restrictedHeaders.add("Host");
+        restrictedHeaders.add("Authorization");
+
+        @SuppressWarnings("unchecked")
+        List<String> headerNames = Collections.list(request.getServletContext().getRequest().getHeaderNames());
+
+        for (String headerName : headerNames) {
+            if (restrictedHeaders.contains(headerName)) {
+                continue;
+            }
+
+            @SuppressWarnings("unchecked")
+            Enumeration<String> headerVals = request.getServletContext().getRequest().getHeaders(headerName);
+            String headerKey = headerName;
+            if (StringUtils.equalsIgnoreCase("X-API-AUTH-HEADER", headerName)) {
+                headerKey = "Authorization";
+            }
+            while (headerVals.hasMoreElements()) {
+                String headerVal = headerVals.nextElement();
+                if (StringUtils.startsWith(headerVal, "rancher:")) {
+                    headerVal = StringUtils.substringAfter(headerVal, "rancher:");
+                }
+                temp.addHeader(headerKey, headerVal);
+            }
+        }
+        temp.addHeader("Host", host);
+
+        if ("POST".equals(method) || "PUT".equals(method)) {
+            InputStream inputStream = request.getInputStream();
+            if (inputStream != null) {
+                temp.bodyByteArray(IOUtils.toByteArray(inputStream));
```

Comment:
You're copying this into memory so the it needs to be limited.  Can you set the default (configurable) limit to 1mb